### ISSUE TITLE
Core does not load filetype icons from theme

### ISSUE THREAD

**Issue by krempelweb** [owncloud/core#15320]
I created a theme today and found out the filetype icons in `/themes/[themename]/core/img/filetypes` will be ignored. As far as I could debug this, I think the bug is located in the function `mimetypeIcon($mimetype)` in `/lib/private/helper.php`.

It only uses `OC::$WEBROOT` in a simple string concat instead of an URL-Generator like `imagePath($app, $image)`.

I don't contribute to the project yet and it's too late now so I cannot fix and test it soon, but perhaps anyone else here can do it.

**Comment by DeepDiver1975** [comment]
afaik a duplicate - @jancborchardt

**Comment by FunkyM** [comment]
This patch adds (basic) support for filetypes within custom themes. However, I think that it is quite inefficient to throw a bunch of file_exists() at the server like this... I think the overall theme handling needs some refactoring to a more static way of handling theme files. Asides this does not fix related issue #6773 as the theme name is not known to the Javascript URL generator causing the stock folder images still to be shown on drag'n'drop despite this patch. Let's call it a start...
```
diff --git a/lib/private/helper.php b/lib/private/helper.php
index cb95dd4..ba0b524 100644
--- a/lib/private/helper.php
+++ b/lib/private/helper.php
@@ -270,32 +270,32 @@ class OC_Helper {
 
 		// Is it a dir?
 		if ($mimetype === 'dir') {
-			self::$mimetypeIcons[$mimetype] = OC::$WEBROOT . '/core/img/filetypes/folder.png';
-			return OC::$WEBROOT . '/core/img/filetypes/folder.png';
+			self::$mimetypeIcons[$mimetype] = self::imagePath('core', 'filetypes/folder.png');
+			return self::$mimetypeIcons[$mimetype];
 		}
 		if ($mimetype === 'dir-shared') {
-			self::$mimetypeIcons[$mimetype] = OC::$WEBROOT . '/core/img/filetypes/folder-shared.png';
-			return OC::$WEBROOT . '/core/img/filetypes/folder-shared.png';
+			self::$mimetypeIcons[$mimetype] = self::imagePath('core', 'filetypes/folder-shared.png');
+			return self::$mimetypeIcons[$mimetype];
 		}
 		if ($mimetype === 'dir-external') {
-			self::$mimetypeIcons[$mimetype] = OC::$WEBROOT . '/core/img/filetypes/folder-external.png';
-			return OC::$WEBROOT . '/core/img/filetypes/folder-external.png';
+			self::$mimetypeIcons[$mimetype] = self::imagePath('core', 'filetypes/folder-external.png');
+			return self::$mimetypeIcons[$mimetype];
 		}
 
 		// Icon exists?
-		if (file_exists(OC::$SERVERROOT . '/core/img/filetypes/' . $icon . '.png')) {
-			self::$mimetypeIcons[$mimetype] = OC::$WEBROOT . '/core/img/filetypes/' . $icon . '.png';
-			return OC::$WEBROOT . '/core/img/filetypes/' . $icon . '.png';
-		}
-
-		// Try only the first part of the filetype
-		$mimePart = substr($icon, 0, strpos($icon, '-'));
-		if (file_exists(OC::$SERVERROOT . '/core/img/filetypes/' . $mimePart . '.png')) {
-			self::$mimetypeIcons[$mimetype] = OC::$WEBROOT . '/core/img/filetypes/' . $mimePart . '.png';
-			return OC::$WEBROOT . '/core/img/filetypes/' . $mimePart . '.png';
-		} else {
-			self::$mimetypeIcons[$mimetype] = OC::$WEBROOT . '/core/img/filetypes/file.png';
-			return OC::$WEBROOT . '/core/img/filetypes/file.png';
+		try {
+			self::$mimetypeIcons[$mimetype] = self::imagePath('core', 'filetypes/' . $icon . '.png');
+			return self::$mimetypeIcons[$mimetype];
+		} catch (RuntimeException $e) {
+			// Try only the first part of the filetype
+			$mimePart = substr($icon, 0, strpos($icon, '-'));
+			try {
+				self::$mimetypeIcons[$mimetype] = self::imagePath('core', 'filetypes/' . $mimePart . '.png');
+				return self::$mimetypeIcons[$mimetype];
+			} catch (RuntimeException $e) {
+				self::$mimetypeIcons[$mimetype] = self::imagePath('core', 'filetypes/file.png');
+				return self::$mimetypeIcons[$mimetype];
+			}
 		}
 	}
 
```

**Comment by jancborchardt** [comment]
It’s related to https://github.com/owncloud/core/issues/6773

@FunkyM can you prepare a pull request for that?

**Comment by MorrisJobke** [comment]
Fix is in #15586

**Comment by MorrisJobke** [comment]
This is a duplicate of #6773

**Comment by MorrisJobke** [comment]
Ah ... #6773 is for the JS part -> reopening

**Comment by MorrisJobke** [comment]
Fixes with #15586
